### PR TITLE
refactor: move contacts out of the wallet

### PR DESCRIPTION
**Pull request by brianp**
Description
---
Refactor contacts and the contact service out of the wallet crate and into it's own crate to be included anywhere needed.

Motivation and Context
---
Wallet doesn't need to manage contacts, and we'll want contacts for other applications so having it as it's own crate makes it portable.

How Has This Been Tested?
---
~Entirely untested. I haven't run this code, only trusting the compiler. Do not merge.~

Tested manually by running the wallet. Many failing tests caught and corrected via CI.

### CODE REVIEW

File: base_layer/contacts/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,73 @@
+[package]
+name = "tari_contacts"
+authors = ["The Tari Development Community"]
+description = "Tari contacts library"
+license = "BSD-3-Clause"
+version = "0.47.0-pre.0"
```
Comment (stringhandler) [inline base_layer/contacts/Cargo.toml]: Please add this file to `scripts/update_crate_metadata.sh`

**Review by stringhandler — COMMENTED** [review]
Looking good so far

File: base_layer/contacts/src/lib.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020, The Tari Project
+// Copyright 2023. The Tari Project
```
Comment (brianp) [inline base_layer/contacts/src/lib.rs]:
```suggestion
// Copyright 2023, The Tari Project
```

File: base_layer/wallet_ffi/wallet.h [hunk 3]
```diff
@@ -68,10 +68,6 @@ struct CommitmentAndPublicKeySignature_RistrettoPublicKey__RistrettoSecretKey;
 
 struct CompletedTransaction;
 
-struct Contact;
-
-struct ContactsLivenessData;
-
```
Comment (brianp) [inline base_layer/wallet_ffi/wallet.h:74]: I feel like this is a failed merge/conflict.

File: base_layer/wallet_ffi/wallet.h [hunk 4]
```diff
@@ -302,9 +298,9 @@ typedef struct Covenant TariCovenant;
 
 typedef struct EncryptedValue TariEncryptedValue;
 
-typedef struct Contact TariContact;
+typedef Contact TariContact;
 
-typedef struct ContactsLivenessData TariContactsLivenessData;
+typedef ContactsLivenessData TariContactsLivenessData;
```
Comment (brianp) [inline base_layer/wallet_ffi/wallet.h:303]: Same. This looks like a conflict.

File: base_layer/wallet_ffi/wallet.h (hunk 3 above)
Comment (brianp) [reply base_layer/wallet_ffi/wallet.h:74]: hmmm this happens on compile, so I guess not a conflict.

**Comment by brianp** [comment]
@stringhandler this is _finally_ ready for a real review.

**Review by sdbondi — APPROVED** [review]
utACK

Nice and neat :)

**Review by SWvheerden — APPROVED** [review]
utACK

**Comment by SWvheerden** [comment]
One thing we might want to decide, for the next PR. 
Is what do we do about the address? Currently, its all tied to the wallet via a `TariAddress`. This means the contacts are all tied per network. Perhaps this is what we want, or perhaps we need to change this to `public key` or both?

File: common_sqlite/src/error.rs [hunk 5]
```diff
@@ -22,12 +22,45 @@
...
+    #[error("Database error: `{0}`")]
+    DieselConnectionError(#[from] diesel::ConnectionError),
+    #[error("Database error: `{0}`")]
+    DatabaseMigrationError(String),
+    #[error("Database error: `{0}`")]
+    BlockingTaskSpawnError(String),
+}
+
+#[derive(Debug, Error)]
+pub enum StorageError {
```
Comment (stringhandler) [inline common_sqlite/src/error.rs:51]: At some point we can merge this and the above. The diesel errors are too general to have as "StorageError"

**Review by stringhandler — APPROVED** [review]

**Pull request merged**